src/haddock/gear/known_cns_errors.py

```diff
@@ -0,0 +1,173 @@
+        " in the relevant CNS scripts."
+        ),
+    "SELRPN error encountered: parsing error": (
+        "Check your restraint files."
+        ),
+    "PARSER error encountered: Encountered too many parsing errors": (
+        "Encountered too many parsing errors. "
+        "Check your input molecules and symmetry restraints."
+        ),
+    "XMREAD error encountered:  sectioning of map incompatible with resolution": (  # noqa : E501
+        "Check your EM map resolution and sectioning."
+        ),
+    "ALLHP error encountered: not enough memory available": (
+        "Too many distance restraints defined. "
+        "Try to reduce this number by checking your definition of active and "
+        "passive residues. "
+        "Make sure to filter those for solvent accessibility. "
+        "Try to decrease the size of your system where possible."
+        ),
+    "error encountered: missing SCATter definition for SELEcted atoms": (
+        "Unsupported atoms/molecules for cryo-EM restraints."
+        ),
+    "ROTMAT error encountered: rotation vector has zero length": (
+        "Check your input parameters and restraints. "
+        "Possibly try turning off the sampling of 180 degrees rotation."
+        )
+    }
+
+
+def find_cns_errors(cns_out_fpath: FilePath) -> Optional[KnownCNSError]:
+    """Detect if a known CNS error is in a cns.out file.
+
+    Parameters
+    ----------
+    cns_out_fpath : FilePath -> Union[str, Path]
+        Path to the cns.out file to check.
+
+    Returns
+    -------
+    Optional[KnownCNSError]
+        An exception for known CNS errors, with its hint on how to solve it!
+    """
+    # Check for file extension
+    if Path(cns_out_fpath).suffix == ".gz":
+        file_handle = gzip.open(cns_out_fpath, "rb")
+    else:
+        file_handle = open(cns_out_fpath, "rb")
+    try:
+        _find_cns_errors(file_handle, KNOWN_ERRORS, filepath=cns_out_fpath)
+    except KnownCNSError as err:
+        return err
+    else:
+        return None
+
+
+def _find_cns_errors(
+        file_handle: Union[gzip.GzipFile, BufferedReader],
+        known_errors: dict[str, str],
+        chunk_size: int = 4096,
+        filepath: FilePath = "",
+        ) -> None:
+    """Backward reading and detect first known CNS error in file.
+
+    Parameters
+    ----------
+    file_handle: Union[gzip.GzipFile, BufferedReader]
+        An opened file in read bytes mode.
+    known_errors : dict[str, str]
+        Dict of known errors and their hints
+    chunk_size : int, optional
+        Check size (in bytes) to read the file backwards, by default 4096
+    filepath : FilePath -> Union[str, Path]
+        Path to the cns.out file currently checked.
+
+    Raises
+    ------
+    KnownCNSError
+        An exception for known CNS errors, with its hint on how to solve it!
+    """
+    # Find file size
+    file_handle.seek(0, 2)
+    size = file_handle.tell()
+    buffer = b''
+    parsed_lines = 99999
```

Comment:
If I understand, this number here sets the max number of lines that will be read, right? Maybe add a comment here to make it clear for prosperity

Reply:
Note that the errors are always appearing at the end of the out files. Anyway to specify to only read the last e.g. 1000 lines?

Reply:
The code is reading the file backwards and stopping when there's an error - up to a upper limit `parsed_lines = 99999`

Reply:
Perfect then!